### PR TITLE
Interpolation function extended to work with FE_Bernstein elements

### PR DESCRIPTION
**Pull request by anates**
Interpolation function extended to also work with FE_Bernstein elements. Compilation fails currently due to has_support_points() not available for DoFHandler::hp-elements. How could I solve that? I tested the SFINAE-approach, but failed until now.

### CODE REVIEW

**Comment by anates** [comment]
Tests fail with (as example from vector_tools/integrate_difference_04): ```
CMakeFiles/integrate_difference_04_hp_02.release.dir/integrate_difference_04_hp_02.cc.o: in function `void test<3>(dealii::VectorTools::NormType, double, double)':
10807: integrate_difference_04_hp_02.cc:(.text._Z4testILi3EEvN6dealii11VectorTools8NormTypeEdd[_Z4testILi3EEvN6dealii11VectorTools8NormTypeEdd]+0x299e): undefined reference to `void dealii::VectorTools::interpolate<3, 3, dealii::TrilinosWrappers::MPI::Vector, dealii::hp::DoFHandler>(dealii::hp::DoFHandler<3, 3> const&, dealii::Function<3, dealii::TrilinosWrappers::MPI::Vector::value_type> const&, dealii::TrilinosWrappers::MPI::Vector&, dealii::ComponentMask const&)
```
Might be related to template specialization I used.

**Comment by anates** [comment]
I tried to improve the code by using partial template specialization, but currently the compiler does not pick up the specialized templates. If you have some suggestions about improvements, I am happy to implement them. Thanks!

**Comment by masterleinad** [comment]
I didn't look at your approach closely, yet. However, you can't partially specialize function templates. If you can, prefer overloading, i.e. split the templated function 
```
    template <int dim,
              int spacedim,
              typename VectorType,
              template <int, int> class DoFHandlerType,
              typename T>
    void
    interpolate(const Mapping<dim, spacedim> &       mapping,
                const DoFHandlerType<dim, spacedim> &dof_handler,
                T &                                  function,
                VectorType &                         vec,
                const ComponentMask &                component_mask)
```
into two templated funtions
```
 template <int dim, int spacedim, typename VectorType>
  void
  interpolate(
    const hp::DoFHandler<dim, spacedim> &                      dof,
    const Function<spacedim, typename VectorType::value_type> &function,
    VectorType &                                               vec,
    const ComponentMask &                                      component_mask)
```
and 
```
 template <int dim, int spacedim, typename VectorType>
  void
  interpolate(
    const DoFHandler<dim, spacedim> &                          dof,
    const Function<spacedim, typename VectorType::value_type> &function,
    VectorType &                                               vec,
    const ComponentMask &                                      component_mask)
```
There should also be declarations in `include/deal.II/numerics/vector_tools.templates.h`. Make sure that these functions are then instantiated in `source/numerics/vector_tools_interpolate.inst.in`.

You also want to run the indentation script (via `make indent`) at some point.

**Comment by anates** [comment]
@masterleinad Thanks for the help! The current status works for FE_Bernstein-elements as well as for FE_Q-elements, but not for FESystem (even though I am not sure if that should work anyway).

**Comment by masterleinad** [comment]
Can you describe the idea in this pull request?

**Comment by anates** [comment]
The idea of this pull request is based on the discussion here: https://groups.google.com/forum/#!searchin/dealii/author$3Ame%7Csort:date/dealii/H5InctUUdg8/zbVOLTR-AgAJ. In short, it aims to making the function VectorTools::interpolate() usable not only for FE_Q-elements, but also for FE_Bernstein-elements. This is done using a similar approach as in VectorTools::get_position_vector(). Does that make sense?

**Comment by masterleinad** [comment]
My question was more about the implementation. So the idea is to interpolate using `FE-Q` elements first in case the finite element is not interpolatory. Then, a transfer matrix is constructed that interpolates the `FE_Q` basis functions in `FE_Q`'s quadrature points. This information is used to create a finite element vector that is interpolatory with respect to `FE_Q`'s quadrature points.

For this to work properly we are assuming that the number of degrees of freedom for the given `FiniteElement` object coincide with the ones for `FE_Q` (for the same degree). In case the underlying polynomial space is the same, the above approach gives a finite element vector that represents the same function as the `FE_Q` interpolation.

**Comment by anates** [comment]
That was the idea, yes. Would that work?

**Comment by masterleinad** [comment]
At the very least, it needs to be properly documented.
Is there any reason why one would want to interpolate using a non-interpolatory finite element instead of using `VectorTools::project`?

**Comment by anates** [comment]
In my project I used interpolate() to interpolate a given function onto the grid while using FE_Q-elements, and wanted to switch to FE_Bernstein-elements without having to rewrite a large part of my code. Thus I tried to extend the interpolate-function as described here.

**Comment by masterleinad** [comment]
OK, thanks for improving the behavior!
Let's go with a good description of what this function is doing for non-interpolatory finite elements then.
In particular, we need to document that this feature is only implemented for `MappingQ1` and not for `hp::DoFhandler`.

Other than that, we also need a test case(have a look [here](https://www.dealii.org/current/developers/testsuite.html#layoutaddtests) and a change log entry in `doc/news/changes/minor`.

It would be great, too, if you could revert the changes to the comments that are not related to this pull request.

**Comment by anates** [comment]
Which comments do you mean? The ones in vector_tools.templates.h? And where should I add the description (in which file)? Thanks!

File: include/deal.II/numerics/vector_tools.templates.h [hunk 1]
```diff
@@ -588,7 +583,118 @@ namespace VectorTools
...
+                    component_mask);
+      }
+    else
+      {
+        AssertDimension(vec.size(), dof.n_dofs());
+        const FiniteElement<dim, spacedim> &fe = dof.get_fe();
+
+        // Construct default fe_mask;
+        const ComponentMask fe_mask(
+          ComponentMask(fe.get_nonzero_components(0).size(), true));
```
Comment (masterleinad) [inline include/deal.II/numerics/vector_tools.templates.h]: Any reason why this can't be
```suggestion
fe.get_nonzero_components(0).size(), true);
```
?

**Comment by anates** [comment]
Will take a look at the comments, but currently I am rather busy with other things, thus it could take a while, if that is ok?

**Comment by masterleinad** [comment]
> Will take a look at the comments, but currently I am rather busy with other things, thus it could take a while, if that is ok?

As long as you will do it eventually, that's totally fine.

File: include/deal.II/numerics/vector_tools.templates.h (hunk 1 above)
Comment (anates) [reply include/deal.II/numerics/vector_tools.templates.h]: Fixed, looks like an artifact from old code

File: source/numerics/vector_tools_interpolate.inst.in [hunk 2]
```diff
@@ -69,16 +69,18 @@ for (VEC : VECTOR_TYPES; deal_II_dimension : DIMENSIONS;
...
-                  const hp::DoFHandler<deal_II_dimension> &,
-                  const Function<deal_II_dimension, VEC::value_type> &,
-                  VEC &,
-                  const ComponentMask &);
+      interpolate(
+        const Mapping<deal_II_dimension> &,
+        const hp::DoFHandler<deal_II_dimension, deal_II_space_dimension> &,
+        const Function<deal_II_dimension, VEC::value_type> &,
+        VEC &,
+        const ComponentMask &);
```
[the remark this answers is not in the document]
Comment (anates) [reply source/numerics/vector_tools_interpolate.inst.in]: Fixed

File: include/deal.II/numerics/vector_tools.templates.h [hunk 3]
```diff
@@ -1074,10 +1180,11 @@ namespace VectorTools
       Vector<number>       tmp(mass_matrix.n());
 
       // If the constraints object does not conflict with the given boundary
-      // values (i.e., it either does not contain boundary values or it contains
-      // the same as boundary_values), we can let it call
+      // values (i.e., it either does not contain boundary values or it
+      // contains the same as boundary_values), we can let it call
```
[the remark this answers is not in the document]
Comment (anates) [reply include/deal.II/numerics/vector_tools.templates.h]: Fixed

File: include/deal.II/numerics/vector_tools.h [hunk 4]
```diff
@@ -622,15 +622,26 @@ namespace VectorTools
...
+    const hp::DoFHandler<dim, spacedim> &                      dof,
+    const Function<spacedim, typename VectorType::value_type> &function,
+    VectorType &                                               vec,
+    const ComponentMask &component_mask = ComponentMask());
+
+  /**
+   * Call the @p interpolate() function above with
+   * <tt>mapping=MappingQGeneric1@<dim>@()</tt> for
+   * <tt>DoFHandlerType=DoFHandler@<dim>@()</tt>.
+   */
```
[the remark this answers is not in the document]
Comment (anates) [reply include/deal.II/numerics/vector_tools.h]: Will work on that later

File: include/deal.II/numerics/vector_tools.templates.h [hunk 5]
```diff
@@ -588,7 +583,119 @@ namespace VectorTools
...
+      {
+        AssertDimension(vec.size(), dof.n_dofs());
+        const FiniteElement<dim, spacedim> &fe = dof.get_fe();
+
+        // Construct default fe_mask;
+        const ComponentMask fe_mask(fe.get_nonzero_components(0).size(), true);
+
+        AssertDimension(
+          fe_mask.size(),
+          fe.get_nonzero_components(0).size()); // A bit redundant, I think?
```
Comment (masterleinad) [inline include/deal.II/numerics/vector_tools.templates.h]: Just remove it.

**Comment by anates** [comment]
> Other than that, we also need a test case(have a look [here](https://www.dealii.org/current/developers/testsuite.html#layoutaddtests) and a change log entry in `doc/news/changes/minor`.
> 
Would it be enough to add tests similar to the current interpolation tests for FE_Q?

**Comment by masterleinad** [comment]
> Would it be enough to add tests similar to the current interpolation tests for FE_Q?

Sure, that is sufficient.

**Comment by anates** [comment]
> > Would it be enough to add tests similar to the current interpolation tests for FE_Q?
> 
> Sure, that is sufficient.

If I understand the math correctly, I should get approximately the same result, regardless if I use FE_Q-elements or FE_Bernstein-elements. Is that correct? If yes, then I can add tests accordingly.

**Comment by masterleinad** [comment]
> If I understand the math correctly, I should get approximately the same result, regardless if I use FE_Q-elements or FE_Bernstein-elements. Is that correct? If yes, then I can add tests accordingly.

That depends on what the test checks. The resulting vector should represent the same interpolation, but its entries will be different from the ones for `FE_Q`.

**Comment by anates** [comment]
> That depends on what the test checks. The resulting vector should represent the same interpolation, but its entries will be different from the ones for `FE_Q`.

How exactly will the entries look different? Unfortunately I am lacking some of the math in the background, thus I am asking

**Comment by masterleinad** [comment]
> How exactly will the entries look different? Unfortunately I am lacking some of the math in the background, thus I am asking

If \phi_i, i=1,...,n are the `FE_Q` base functions, and \psi_i, i=1,...n the `FE_Bernstein` base functions, we expect `VectorTools::interpolate` to provide coefficients c_i, resp. d_i, i=1,...,n such that \sum_i c_i \phi_i = \sum_i d_i \psi_i (in every point). Since the basis functions are different, I would expect c_i and d_i to be different as well.

**Comment by bangerth** [comment]
@anates -- any news?

**Comment by anates** [comment]
Currently the aim of my main project has moved, thus I had to put this request aside until I have more time again, unfortunately. Will keep it in mind, though.

**Comment by bangerth** [comment]
OK, let us know when you want us to look at this again. Should we close the pull request for the moment? It can always be re-opened once you're ready to work on it again!

**Comment by anates** [comment]
We can do that, I will not have time for at least the next two months to work on that PR, unfortunately.

**Pull request closed without merging**